File: platform/platform-impl/src/com/intellij/openapi/application/ApplicationStarterBase.java
// Copyright 2000-2022 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package com.intellij.openapi.application;

import com.intellij.ide.CliResult;
import com.intellij.idea.SocketLock;
import com.intellij.openapi.editor.Document;
import com.intellij.openapi.fileEditor.FileDocumentManager;
import com.intellij.openapi.ui.Messages;
import com.intellij.openapi.util.NlsContexts;
import com.intellij.openapi.vfs.VirtualFile;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Arrays;
import java.util.List;
import java.util.concurrent.Future;

/**
 * @author Konstantin Bulenkov
 */
@SuppressWarnings({"UseOfSystemOutOrSystemErr", "CallToPrintStackTrace"})
public abstract class ApplicationStarterBase implements ApplicationStarter {
  private final int[] myArgsCount;

  protected ApplicationStarterBase(int... possibleArgumentsCount) {
    myArgsCount = possibleArgumentsCount;
  }

  @Override
  public boolean isHeadless() {
    return false;
  }

  @Override
  public boolean canProcessExternalCommandLine() {
    return true;
  }

  @NotNull
  @Override
  public Future<CliResult> processExternalCommandLineAsync(@NotNull List<String> args, @Nullable String currentDirectory) {
    if (!checkArguments(args)) {
      String title = ApplicationBundle.message("app.command.exec.error.title", getCommandName());
      ApplicationManager.getApplication().invokeLater(() -> Messages.showMessageDialog(getUsageMessage(), title, Messages.getInformationIcon()));
      return CliResult.error(1, getUsageMessage());
    }
    try {
      return processCommand(args, currentDirectory);
    }
    catch (Exception e) {
      String title = ApplicationBundle.message("app.command.exec.error.title", getCommandName());
      String message = ApplicationBundle.message("app.command.exec.error", getCommandName(), e.getMessage());
      ApplicationManager.getApplication().invokeLater(() -> Messages.showMessageDialog(message, title, Messages.getErrorIcon()));
      return CliResult.error(1, message);
    }
  }

  protected static void saveAll() {
    FileDocumentManager.getInstance().saveAllDocuments();
    ApplicationManager.getApplication().saveSettings();
  }

  protected static void saveIfNeeded(@Nullable VirtualFile file) {
    if (file == null) return;
    Document document = FileDocumentManager.getInstance().getCachedDocument(file);
    if (document != null) FileDocumentManager.getInstance().saveDocument(document);
  }

  protected boolean checkArguments(@NotNull List<String> args) {
    return Arrays.binarySearch(myArgsCount, args.size() - 1) != -1 && getCommandName().equals(args.get(0));
  }

  @NlsContexts.DialogMessage
  public abstract String getUsageMessage();

  @NotNull
  protected abstract Future<CliResult> processCommand(@NotNull List<String> args, @Nullable String currentDirectory) throws Exception;

  @Override
  public void premain(@NotNull List<String> args) {
    if (!checkArguments(args)) {
      System.err.println(getUsageMessage());
      System.exit(1);
    }
  }

  @Override
  public void main(@NotNull List<String> args) {
    try {
      int exitCode;
      try {
        String currentDirectory = System.getenv(SocketLock.LAUNCHER_INITIAL_DIRECTORY_ENV_VAR);
        Future<CliResult> commandFuture = processCommand(args, currentDirectory);
        CliResult result = commandFuture.get();
        if (result.message != null) {
          System.out.println(result.message);
        }
        exitCode = result.exitCode;
      }
      finally {
        ApplicationManager.getApplication().invokeAndWait(ApplicationStarterBase::saveAll);
      }
      System.exit(exitCode);
    }
    catch (Exception e) {
      e.printStackTrace();
      System.exit(1);
    }
    catch (Throwable t) {
      t.printStackTrace();
      System.exit(2);
    }
  }
}

File: platform/platform-impl/src/com/intellij/openapi/application/ApplicationStarterBase.kt
// Copyright 2000-2023 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package com.intellij.openapi.application

import com.intellij.configurationStore.saveSettings
import com.intellij.ide.CliResult
import com.intellij.openapi.fileEditor.FileDocumentManager
import com.intellij.openapi.ui.Messages
import com.intellij.openapi.util.NlsContexts
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.platform.ide.bootstrap.LAUNCHER_INITIAL_DIRECTORY_ENV_VAR
import com.intellij.platform.ide.bootstrap.commandNameFromExtension
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.jetbrains.annotations.ApiStatus.Internal
import java.util.*
import kotlin.system.exitProcess

/**
 * @author Konstantin Bulenkov
 */
@Internal
abstract class ApplicationStarterBase protected constructor(private vararg val argsCount: Int) : ModernApplicationStarter() {
  abstract val usageMessage: @NlsContexts.DialogMessage String?

  override val isHeadless: Boolean
    get() = false

  companion object {
    @JvmStatic
    protected fun saveIfNeeded(file: VirtualFile?) {
      if (file == null) {
        return
      }
      FileDocumentManager.getInstance().getCachedDocument(file)?.let(FileDocumentManager.getInstance()::saveDocument)
    }
  }

  override fun canProcessExternalCommandLine(): Boolean = true

  override suspend fun processExternalCommandLine(args: List<String>, currentDirectory: String?): CliResult {
    val commandName = commandNameFromExtension
    if (!checkArguments(args)) {
      val title = ApplicationBundle.message("app.command.exec.error.title", commandName)
      withContext(Dispatchers.EDT) {
        Messages.showMessageDialog(usageMessage, title, Messages.getInformationIcon())
      }
      return CliResult(1, usageMessage)
    }

    try {
      return executeCommand(args, currentDirectory)
    }
    catch (e: CancellationException) {
      throw e
    }
    catch (e: Exception) {
      e.printStackTrace() // The dialog may sometimes not be shown, e.g., in remote dev scenarios.
      val title = ApplicationBundle.message("app.command.exec.error.title", commandName)
      val message = ApplicationBundle.message("app.command.exec.error", commandName, e.message)
      withContext(Dispatchers.EDT) {
        Messages.showMessageDialog(message, title, Messages.getErrorIcon())
      }
      return CliResult(1, message)
    }
  }

  protected open fun checkArguments(args: List<String>): Boolean {
    return Arrays.binarySearch(argsCount, args.size - 1) >= 0 && commandNameFromExtension == args[0]
  }

  protected abstract suspend fun executeCommand(args: List<String>, currentDirectory: String?): CliResult

  override fun premain(args: List<String>) {
    if (!checkArguments(args)) {
      System.err.println(usageMessage)
      exitProcess(1)
    }
  }

  final override suspend fun start(args: List<String>) {
    try {
      val exitCode: Int = try {
        val result = executeCommand(args = args, currentDirectory = System.getenv(LAUNCHER_INITIAL_DIRECTORY_ENV_VAR))
        result.message?.let(::println)
        result.exitCode
      }
      finally {
        withContext(Dispatchers.EDT) {
          writeIntentReadAction {
            FileDocumentManager.getInstance().saveAllDocuments()
          }
        }
        saveSettings(ApplicationManager.getApplication())
      }
      exitProcess(exitCode)
    }
    catch (e: CancellationException) {
      throw e
    }
    catch (e: Exception) {
      e.printStackTrace()
      exitProcess(1)
    }
    catch (t: Throwable) {
      t.printStackTrace()
      exitProcess(2)
    }
  }
}
